rfd/0203-database-healthchecks.md

```diff
@@ -0,0 +1,611 @@
+This special case is to bound the amount of time spent in the `initialized` status if health checks flap between pass/fail without reaching either threshold.
+
+### Types of health checks
+
+For the initial implementation we will only support TCP health checks and only for databases.
+
+A TCP health check is generic and simple, so the initial implementation will support health checks for all databases that Teleport supports: Postgres, MySQL, MongoDB, etc.
+
+We can extend this feature later to include additional health check protocols such as HTTP and TLS checks.
+
+We can also extend this feature to other one-agent-to-many-resources types of resources: apps, Windows desktops, OpenSSH servers, etc.
+
+#### TCP health check
+
+A TCP health attempts to establish a TCP connection to the target address.
+If dialing the target address times out or receives a TCP reset (RST), then the check has failed.
+Otherwise, the check has passed.
+
+### Configuration
+
+Health checks will be an opt-out configurable setting with reasonable defaults.
+
+A new cluster resource will be added to manage health check settings: `health_check_config`.
+
+A `health_check_config` preset resource named "default" will be added.
+
+The "default" `health_check_config` preset will enable health checks for all databases.
+
+Users will be able to easily opt-out of health checks by setting `enabled: false` on the "default" `health_check_config`.
+
+Teleport database agents will cache `health_check_config` resources.
+
+#### Configuration settings
+
+The `health_check_config` resource will expose the following settings:
+
+- Match: Resources with matching labels will use these health check settings.
+- Enabled: whether to enable health checks
+- Interval: time between each health check
+- Timeout: the health check connection attempt timeout (timing out fails the check)
+- Healthy Threshold: number of consecutive passing health checks after which the resource health status is changed to `healthy`
+- Unhealthy Threshold: number of consecutive failing health checks after which the resource health status is changed to `unhealthy`
+
+These settings are sufficient to support TCP connection health checks in the initial implementation.
+They are also necessary for other types of health check, including TCP, TLS, and HTTP(S).
+Thus, they generalize well.
+Future work can extend `health_check_config` with more specific settings as needed to support new health check types.
+
+#### Configuration label matching 
+
+The settings in a `health_check_config` resource will be used for a database if all of the labels in `health_check_config.spec.match.db_labels` match the database's labels.
+
+If multiple `health_check_config` resources match a database, then the `health_check_config` resources will be sorted by name and only the first will apply.
```

Comment:
I still feel uneasy about using sorted names to decide which health check to apply.

What do you guys think if we introduce the `priority` flag and higher priority config will be applied?

We can ensure that priority is unique when adding/updating the config resource across all configs, and we could set a max limit say 100 to avoid having a huge number of them.

Reply:
Generally I'm in favor of as little new configuration as possible so sorting by name for now seems fine. We can always add priority if needed. I understand it's pretty critical for things like login rules where order of evaluation matters but for health checks feels like it's not as critical.